### PR TITLE
Add e2e-spock-geb quickstarter

### LINKED ISSUE #89: (E2E) Test Quickstarter with Spock/Geb/Unirest for Java and Groovy
**Is your feature request related to a problem? Please describe.**
Once all code-related and service-related components have been deployed (using the release manager), I want to be able to run automated tests against any combination of live components (integration, acceptance and system-level tests).

**Describe the solution you'd like**
The quickstarter-embedded framework shall support writing tests using [Spock](http://spockframework.org), [Geb](https://gebish.org) and [Unirest](http://unirest.io), and provide a test structure for integration and acceptance testing for both Java and Groovy.

**Additional context**
Following the existing pattern, we would call this one *e2e-spock-geb*. As a side note, I find the *e2e* misleading as this name is closest to system-level than acceptance or integration testing. @michaelsauter let me know if you think we should reconsider this prefix.

### PR DESCRIPTION
**Pull request by metmajer**
Closes #89.

**TODO**
- [x] Assert tests can be provided in both Java and Groovy
- [x] Assert stashed test results (xUnit XML documents) for _acceptance_ and _integration_ tests can be restored by the Release Manager
- [ ] Support `installation` tests before `integration` tests
- [x] Add Antora documentation to https://github.com/opendevstack/ods-quickstarters/tree/feature/add-e2e-spock-geb/docs/modules/ROOT
- [ ] ~Optimize bootstrap time by pre-installing dependencies~
- [x] Add component metadata in metadata.yml (see #144)

### CODE REVIEW

**Comment by metmajer** [comment]
@hrcornejo I made some modifications to our previous work:

1) Our quickstarter is now correctly based on the new `ods-quickstarters` project, as opposed to the meanwhile removed `ods-project-quickstarters`.
2) I have removed our (explicit) directory creation and copying of xUnit XML documents to a path supported by the Release Manager in favor of using an `OdsContext::testResults` configuration parameter provided by the `ods-jenkins-shared-library` [here](https://github.com/opendevstack/ods-quickstarters/pull/91/files#diff-6635af705ca56d156c44b4e3d04e73b7R46) (look [here](https://github.com/opendevstack/ods-jenkins-shared-library/blob/c43057a05411b71cef101bb31b629d6f7f851cd1/src/org/ods/OdsPipeline.groovy#L148) for further details).
3) I have added support for a Java property `test.ignoreFailures` to control the ignoring of test failures in the test execution to both _integration_ and _acceptance_ Gradle tasks [here](https://github.com/opendevstack/ods-quickstarters/pull/91/files#diff-0f599e942076f73f203df8fd73849ddcR107-R126). We expose this as a feature to the user when we invoke `gradlew test` in the `Jenkinsfile.template` [here](https://github.com/opendevstack/ods-quickstarters/pull/91/files#diff-6635af705ca56d156c44b4e3d04e73b7R49).
4) I have added an `.editorconfig` into the `files` directory of this quickstarter.

**Comment by metmajer** [comment]
@hrcornejo Can you please verify by creating a simple test in each of the following directories that this quickstarter supports tests written in both `Java` and `Groovy`?

- `test/acceptance/java` and `test/acceptance/groovy`
- `test/integration/java` and `test/integration/groovy`

Looking at `build.gradle` [here](https://github.com/opendevstack/ods-quickstarters/pull/91/files#diff-0f599e942076f73f203df8fd73849ddcR77-R79) and [here](https://github.com/opendevstack/ods-quickstarters/pull/91/files#diff-0f599e942076f73f203df8fd73849ddcR88-R90), acceptance and integration test `srcDir`s may only support Groovy at the moment.

**Comment by metmajer** [comment]
@michaelsauter With respect to optimizing bootstrap time, would it be feasible to install package Gradle dependencies (Spock, Geb, JUnit5, etc.) by mimicking what you have done in the `be-java-springboot` quickstarter [here](https://github.com/opendevstack/ods-quickstarters/blob/master/be-java-springboot/Jenkinsfile#L42-L45)?

To be frank, I am not sure how what you do in a quickstarter's `Jenkinsfile` affects the runtime behavior of a component. What I'd like to avoid is having to install all Gradle dependencies at test execution time.

**Comment by michaelsauter** [comment]
@metmajer The linked code does not help with bootstrap time. What you linked is part of the "provisioning process" of a quickstarter - a one time thing. The only way to minimise the runtime of a Jenkins job is to install the dependencies into the Jenkins agent image you use. That means either we add the dependencies into the one you are using (based on `Jenkinsfile.template`, it is the maven image), or you create a separate image specifically for your quickstarter.

**Comment by hrcornejo** [comment]
@metmajer Now the tests can be in:
- test/acceptance/java and test/acceptance/groovy
- test/integration/java and test/integration/groovy

To add support simultaneously JUnit5 and Groovy + Spock have been defined 4 task in gradle to use or not [useJUnitPlatform()](https://github.com/opendevstack/ods-quickstarters/pull/91/files/8f10a87f099e55188e33ffcf5cdd9d533e1b3f55..7884b02dd1fae84d5f7db7e8d907366a0f2625fe#diff-0f599e942076f73f203df8fd73849ddcR121). With useJUnitPlatform only JUnit could be launched, without it only Groovy could be launched.

**Comment by metmajer** [comment]
@hrcornejo are you sure we need these different tasks? Gradle supports building an application composed of Java and Groovy code using `gradle build`, given that an `apply: groovy` has been provided in the Gradle build file. I would assume that `gradle test` makes no difference here.

I would not like us to break with this convenience. I assume this is because of the `integration` and `acceptance` folders deviating from the default? Can you please do a little research?

**Comment by metmajer** [comment]
@hrcornejo please add another test type `integration` and make sure it is invoked before the `integration` tests. Thanks!

**Comment by hrcornejo** [comment]
@metmajer added the new type installation and refactorized the build.config to get only onenew task (parametrized)

File: e2e-spock-geb/Jenkinsfile.template [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  def javaOpts = "-Xmx512m"
+  def gradleTestOpts = "-Xmx128m"
+  def springBootEnv = context.environment
+  if (springBootEnv.contains('-dev')) {
+    springBootEnv = 'dev'
+  }
+
+  stage('Test') {
+    withEnv(["TAGVERSION=${context.tagversion}", "NEXUS_HOST=${context.nexusHost}", "NEXUS_USERNAME=${context.nexusUsername}", "NEXUS_PASSWORD=${context.nexusPassword}", "JAVA_OPTS=${javaOpts}","GRADLE_TEST_OPTS=${gradleTestOpts}","ENVIRONMENT=${springBootEnv}"]) {
+      def status = sh(script: "./gradlew clean test --stacktrace --no-daemon -Dtest.ignoreFailures=true", returnStatus: true)
```
Comment (metmajer) [inline e2e-spock-geb/Jenkinsfile.template]: Let's remove the use of ignoreFailures from these lines and provide a central variable in build.gradle.

File: e2e-spock-geb/files/build.gradle [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    return tasks.create(generateTaskName(type, language), Test) {
+        description = "Runs ${type} tests ${language}."
+        group = "verification"
+
+        if ("${language}".equalsIgnoreCase(TestLanguages.JAVA)) {
+            // Mandatory to run JUnit 5 test but incompatible with groovy tests
+            useJUnitPlatform()
+        }
+
+        ignoreFailures = Boolean.getBoolean("test.ignoreFailures")
```
Comment (metmajer) [inline e2e-spock-geb/files/build.gradle]: @hrcornejo let's set this to `true` by default and add a comment on top what the user can control here.

File: docs/modules/ROOT/pages/e2e-spock-geb.adoc [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+```
+nexus_url=<URL_OF_YOUR_NEXUS_SERVER>
+nexus_user=<YOUR_NEXUS_USERNAME>
+nexus_pw=<YOUR_NEXUS_PASSWORD>
+```
+
+If you do not want to use Nexus at all, just define the following property:
+```
+no_nexus=true
```
[the remark this answers is not in the document]
Comment (clemensutschig) [reply docs/modules/ROOT/pages/e2e-spock-geb.adoc]: @metmajer  - I guess to run local - w/o any access to nexus :)

[the remark this answers is not in the document]
Comment (hrcornejo) [reply docs/modules/ROOT/pages/e2e-spock-geb.adoc]: @clemensutschig @metmajer yes. To run locally

File: e2e-spock-geb/Jenkinsfile.template (hunk 1 above)
Comment (hrcornejo) [reply e2e-spock-geb/Jenkinsfile.template]: Done

File: docs/modules/ROOT/pages/e2e-spock-geb.adoc [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+******* http://spockframework.org/[spock]
+
+******* https://gebish.org/[geb]
+
+******* http://unirest.io/[unirest]
+
+## Usage - how do you start after you provisioned this quickstarter
+
+* Run command `gradlew test -Dtest.ignoreFailures=true` in project directory to execute the end-to-end tests via spock/geb against the Google Home page and demo jUnit 5 tests.
```
[the remark this answers is not in the document]
Comment (hrcornejo) [reply docs/modules/ROOT/pages/e2e-spock-geb.adoc]: Done

File: e2e-spock-geb/files/build.gradle (hunk 2 above)
Comment (hrcornejo) [reply e2e-spock-geb/files/build.gradle]: Done

**Comment by hrcornejo** [comment]
@metmajer with this quickstarter the user can:
- To write tests in Java using JUnit 5 annotations (examples in installation/integration/acceptance - java folders)
- To write tests in Groovy with Spock  (examples in installation/integration/acceptance - groovy folders), not bound to use JUnit 5.

**Comment by metmajer** [comment]
@hrcornejo as we just discussed, please try to make Spock work for tests written in Java, too. Thanks 🥇

**Comment by hrcornejo** [comment]
@metmajer It's not possible to run Spock with Java tests. Spock needs keywords (given, then,....) and Groovy syntax that doesn't work in plain Java, Only in groovy classes.

**Comment by metmajer** [comment]
@clemensutschig @michaelsauter @gerardcl ready for approval

**Comment by clemensutschig** [comment]
Needs to also go into prov App cm then

**Comment by michaelsauter** [comment]
I am wondering about the general strategy here. Given we already have an E2E testing QS (cypress based), where does the need for a Spock based come from? And how do the two solutions compare? I haven't seen any reasoning yet, and I would like to see that - plus I think it would be good to have this somewhere in docs for others wondering the same thing.

Further, I am wondering if this should go into the official repo straight away or if we want to give this some "lab" time?

**Comment by gerardcl** [comment]
Hi! LGTM! I would like to use it! :) 
but! having said that and following thoughts from last Michael's comment:
- is this still strongly related to goals from #145? if yes, I see it having an reason/exception to jump directly into ods-quickstarters, if not:
- we should follow an usage monitoring approach regarding each current quickstarter in ods-quickstarters (which I would foresee some of them already movable to a non full official repo/ lab repo, be named like ods-quickstarters-incubator; like some fe-* quickstarters -which are hard to maintain for example-, or ds-python-dash -which has been recently proposed and seems of high interest/value in the DS context, or any possible incoming be-<lang>-gateway)

regarding last comment, there is already a [joint effort activity](https://gist.github.com/michaelsauter/0e8254403d8eaeefbb07c23f6d2a4bcc#quickstarter-ownership-and-quality-improvements) detecting which current ods-quickstarters could potentially be kept in the main repo and which not

I would like to have this discussion (about this specific quickstarter and the others) before approving! :)

**Comment by metmajer** [comment]
There is as much or little need for Spock as it is for Cypress. Both are similar in what they can achieve, and in the end, it's a matter of choice and taste. Personally, having done more backend than frontend development, I prefer to have a framework where I can write super concise tests, but have access to entire the Java ecosystem, if needed. Spock exists for a decade, is used by big and small companies. It deserves the label enterprise-grade and haven't come across a better solution yet.

Other reasons:

- It is currently the only quickstarter for e2e testing that supports the execution of integration, acceptance and installation tests, and the proper delivery of its results to the release manager for various ensuing reporting tasks.
- In our enterprise environment, we see it as the #1 choice and will provide proper support and enablement.

**Comment by metmajer** [comment]
@clemensutschig @michaelsauter @gerardcl Because of the above rationale, and since it is a center-piece of our upcoming delivery, without which we would otherwise be blocked, I proceed to merge mainline.

**Pull request merged**